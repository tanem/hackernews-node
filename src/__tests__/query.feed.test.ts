import bcrypt from 'bcryptjs'
import Chance from 'chance'
import path from 'path'
import request from 'supertest'

const chance = Chance(path.basename(__filename))

let httpServerUrl: typeof global.__HTTP_SERVER_URL__
let prisma: typeof global.__PRISMA__

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()
const links = new Array(5).fill(undefined).map(() => ({
  description: chance.sentence({ words: 5 }),
  url: chance.url(),
}))

beforeEach(async () => {
  await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
      links: {
        create: links,
      },
    },
  })
})

test('simple', async () => {
  const { body, status } = await request(httpServerUrl).post('/').send({
    query: `
      query {
        feed {
          count
          links {
            id
            description
            url
            postedBy {
              id
              name
            }
            votes {
              id
              user {
                id
                name
              }
            }
          }
        }
      }
    `,
  })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('filter', async () => {
  const { body, status } = await request(httpServerUrl).post('/').send({
    query: `
      query {
        feed(filter: "mo") {
          count
          links {
            id
            description
            url
            postedBy {
              id
              name
            }
            votes {
              id
              user {
                id
                name
              }
            }
          }
        }
      }
    `,
  })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('pagination', async () => {
  const { body, status } = await request(httpServerUrl).post('/').send({
    query: `
      query {
        feed(
          take: 3
          skip: 1
        ) {
          count
          links {
            id
            description
            url
            postedBy {
              id
              name
            }
            votes {
              id
              user {
                id
                name
              }
            }
          }
        }
      }
    `,
  })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('sorting', async () => {
  const { body, status } = await request(httpServerUrl).post('/').send({
    query: `
      query {
        feed(orderBy: { description: asc }) {
          count
          links {
            id
            description
            url
            postedBy {
              id
              name
            }
          }
        }
      }
    `,
  })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})
