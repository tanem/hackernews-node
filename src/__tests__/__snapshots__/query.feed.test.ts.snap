// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter 1`] = `
Object {
  "data": Object {
    "feed": Object {
      "count": 0,
      "links": Array [],
    },
  },
}
`;

exports[`pagination 1`] = `
Object {
  "data": Object {
    "feed": Object {
      "count": 5,
      "links": Array [
        Object {
          "description": "Fef no gocinfan fe gu.",
          "id": "2",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://dum.dj/tofej",
          "votes": Array [],
        },
        Object {
          "description": "Jutezuj ce ova wozo tim.",
          "id": "3",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://ife.fr/rogeuco",
          "votes": Array [],
        },
        Object {
          "description": "Sa uvian ed laurapu re.",
          "id": "4",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://zobekdiv.sg/soki",
          "votes": Array [],
        },
      ],
    },
  },
}
`;

exports[`simple 1`] = `
Object {
  "data": Object {
    "feed": Object {
      "count": 5,
      "links": Array [
        Object {
          "description": "Fano ki wetu lehmuh razi.",
          "id": "1",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://bizbah.yt/mag",
          "votes": Array [],
        },
        Object {
          "description": "Fef no gocinfan fe gu.",
          "id": "2",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://dum.dj/tofej",
          "votes": Array [],
        },
        Object {
          "description": "Jutezuj ce ova wozo tim.",
          "id": "3",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://ife.fr/rogeuco",
          "votes": Array [],
        },
        Object {
          "description": "Sa uvian ed laurapu re.",
          "id": "4",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://zobekdiv.sg/soki",
          "votes": Array [],
        },
        Object {
          "description": "Useri rorur piz ruefuaj hebak.",
          "id": "5",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://hasme.mm/bubtocfab",
          "votes": Array [],
        },
      ],
    },
  },
}
`;

exports[`sorting 1`] = `
Object {
  "data": Object {
    "feed": Object {
      "count": 5,
      "links": Array [
        Object {
          "description": "Fano ki wetu lehmuh razi.",
          "id": "1",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://bizbah.yt/mag",
        },
        Object {
          "description": "Fef no gocinfan fe gu.",
          "id": "2",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://dum.dj/tofej",
        },
        Object {
          "description": "Jutezuj ce ova wozo tim.",
          "id": "3",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://ife.fr/rogeuco",
        },
        Object {
          "description": "Sa uvian ed laurapu re.",
          "id": "4",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://zobekdiv.sg/soki",
        },
        Object {
          "description": "Useri rorur piz ruefuaj hebak.",
          "id": "5",
          "postedBy": Object {
            "id": "1",
            "name": "Johnny",
          },
          "url": "http://hasme.mm/bubtocfab",
        },
      ],
    },
  },
}
`;
